src/Parser/Shaman/Restoration/Modules/Talents/EchoReset.js

```diff
@@ -0,0 +1,44 @@
+import SPELLS from 'common/SPELLS';
+import ITEMS from 'common/ITEMS';
+
+import Analyzer from 'Parser/Core/Analyzer';
+import SpellUsable from 'Parser/Core/Modules/SpellUsable';
+import Combatants from 'Parser/Core/Modules/Combatants';
+
+/*
+* Lazy solution for the Riptide reset, as it is undetectable but the spell is too important to disable.
+* Cast efficiency with echo selected is a lot harsher compared to without to account for the resets.
+*/
+
+class EchoReset extends Analyzer {
+  static dependencies = {
+    spellUsable: SpellUsable,
+    combatants: Combatants,
+  };
+
+  lastRiptideTimestamp = 0;
+  lastRiptideCharges = 0;
+
+  on_initialized(){
+    this.active = this.combatants.selected.hasTalent(SPELLS.ECHO_OF_THE_ELEMENTS_TALENT.id) || this.combatants.selected.hasFinger(ITEMS.SOUL_OF_THE_FARSEER.id);
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (spellId !== SPELLS.RIPTIDE.id) {
+      return;
+    }
+    if (this.spellUsable.chargesAvailable(SPELLS.RIPTIDE.id) === 0) {
+        if(event.timestamp - this.lastRiptideTimestamp <= 6000 && this.lastRiptideCharges <= 1) {
+            this.spellUsable.endCooldown(spellId,false,event.timestamp,6000);
+            this.spellUsable.beginCooldown(spellId);
```

Comment:
Any reason why you don't just use reduceCooldown instead of this and the above line like you do in LavaSurge? This will essentially reset both charges as it first ends the cooldown of the current charge, and then reduce the cooldown of the second charge by 6000 ms.